src/adic_space.lean

```diff
@@ -8,7 +8,63 @@ import for_mathlib.topology
 
 open nat function
 
-variables (R : Type) [comm_ring R] [topological_space R] [topological_ring R]  
+section comm_ring
+variables (R : Type) [comm_ring R]
+-- This section is filled in in Johan's PR
+definition is_subring {R : Type} [comm_ring R] : set R → Prop := sorry 
+definition is_integrally_closed {R : Type} [comm_ring R] : set R → Prop := sorry  
+end comm_ring
+
+section topological_ring
+
+-- The following two lines are a nasty hack to switch between having the carrier 
+-- Type implicit or explicit.
+variables (R : Type) [comm_ring R] [topological_space R] [comm_ring R] [topological_ring R]  
+variables {A : Type} [comm_ring A] [topological_space A] [comm_ring A] [topological_ring A]  
+
+def nhd_zero := (nhds (0 : R)).sets
+
+-- peredicates we need for topological rings
+
+-- Following is copy-pasted from t2_space class.
+-- We need to think whether we could directly use that class
+definition is_hausdorff (α : Type*) [topological_space α] : Prop := 
+  ∀x y, x ≠ y → ∃u v : set α, is_open u ∧ is_open v ∧ x ∈ u ∧ y ∈ v ∧ u ∩ v = ∅
+
+open filter
+instance toplogical_ring.to_uniform_space : uniform_space R := 
+uniform_space.of_core { uniformity  := (⨅ U ∈ nhd_zero R, principal {p : R×R | p.2 - p.1 ∈ U}),
+                        refl        := begin simp, intros i H r, exact mem_of_nhds H end,
+                        symm        := sorry,
+                        comp        := sorry }
+
```

Comment:
Patrick I'm rather relieved you've stepped in here to unravel the uniform spaces and filters. But what are these sorrys?